lib/puppet/util/package/version/pip.rb

```diff
@@ -0,0 +1,230 @@
+module Puppet::Util::Package::Version
+  class Pip
+    def initialize(version)
+      matched = version.match(Regexp.new(("^\\s*") + @@VERSION_PATTERN + ("\\s*$"), Regexp::EXTENDED | Regexp::MULTILINE | Regexp::IGNORECASE))
+
+      raise ValidationFailure.new(version) unless matched
+      self.class.include Comparable
+
+      @epoch_data   = matched[:epoch].to_i
+      @release_data = matched[:release].split('.').map{ |r| r.to_i }                                 if matched[:release]
+      @pre_data     = parse_letter_version(matched[:pre_l], matched[:pre_n])                         if matched[:pre_l]  || matched[:pre_n]
+      @post_data    = parse_letter_version(matched[:post_l], matched[:post_n1] || matched[:post_n2]) if matched[:post_l] || matched[:post_n1] || matched[:post_n2]
+      @dev_data     = parse_letter_version(matched[:dev_l], matched[:dev_n])                         if matched[:dev_l]  || matched[:dev_n]
+      @local_data   = parse_local_version(matched[:local])                                           if matched[:local]
+
+      @key = compose_key(@epoch_data, @release_data, @pre_data, @post_data, @dev_data, @local_data)
+    end
+
+    def self.parse(version)
+      new(version)
+    end
+
+    def self.compare(version_a, version_b)
+      version_a = new(version_a) unless version_a.is_a?(self.class)
+      version_b = new(version_b) unless version_a.is_a?(self.class)
+
+      version_a <=> version_b
+    end
+
+    def to_s
+      parts = []
+
+      parts.push("#{epoch}!")                   if epoch
+      parts.push(release.join("."))             if release
+      parts.push(pre_release)                   if pre_release
+      parts.push(".post#{post_release[1]}")     if post_release
+      parts.push(".dev#{dev_release[1]}")       if dev_release
+      parts.push("+#{local_version.join(".")}") if local_version
+
+      parts.join
+    end
+
+    def <=>(other)
+      raise ValidationFailure.new(other.to_s) unless other.is_a?(self.class)
+      compare(self.key, other.key)
+    end
+
+    def base_version
+      to_s.split('+').first
+    end
+
+    def major_version
+      release[0].to_i
+    end
+
+    def minor_version
+      release[1].to_i
+    end
+
+    def micro_version
+      release[2].to_i
+    end
+
+    def is_pre_release?
+      dev_release != nil || pre_release != nil
+    end
+
+    def is_post_release?
+      post_release != nil
+    end
+
+    def is_dev_release?
+      dev_release != nil
+    end
+
+    def key
+      @key
+    end
+
+    private
+
+    @key = nil
+
+    def key=(val)
+      @key = val
+    end
+
+    def epoch
+      @epoch_data if @epoch_data != 0 && @epoch_data
```

Comment:
i would check first for `@epoch_data`, it makes more sense to check for existence before checking for specific value